### PR TITLE
[CDU] Fix error message deletes input, add support for input delay

### LINKED ISSUE #1754: [BUG] [MCDU] System message should not delete pilot input
**Mod Version**

4th Nov 20, dev/master version

**Describe the bug**

When MCDU scratchpad contains a pilot input which get overritten by a system message like "NOT ALLOWED", clearing it should not delete the pilot input.

**To Reproduce**

1. Load the plane. Go into eg. INIT page
2. Write eg. "ABC" into the scratchpad and click on CRZ FL LSK
3. "NOT ALLOWED" will popup
4. delete with CLR key
5. scratchpad is emtpy
You can reproduce it everywhere a system message gets loaded

**Expected behavior**

After clearing the system message, the pilot input should persist in the cratchpad

**Actual behavior**

pilot input gets deleted

**References**

**Input "ABC"**
image

**Click CRZ FL**
image

**CLR**
image

**Additional context**

Was this working before/when did the issue start occurring?

Is this a problem in the vanilla unmodded game?

Discord username (if different from GitHub):

### PR DESCRIPTION
**Pull request by MisterChocker**
Fixes #1754

## Summary of Changes

- centralized input handling
   - fixes input being deleted on error massage
   - supports input delay
   - updated cdu/atsu/options pages to work with changed input handling
- made fake error messages proper error massages
- added input delay for
   - menu buttons
   - 1-6L and 1-6R buttons
      - added option to override/change delay individual
      - added overrides for switch page actions to all files

I have categorized buttons in two types:
- Buttons who switch pages _(using `getDelaySwitchPage()`)_
- Buttons who enter data _(using `getDelayBasic()`)_

Buttons 1-6L and 1-6R are using by default `getDelayBasic()`.

By placing either `mcdu.leftInputDelay` or `mcdu.rightInputDelay`, with the corresponding index behind, above the input handler, an override will be created. The override returns a number. I recommend using or creating centralized methods for this.

```js
mcdu.leftInputDelay[1] = () => {
    return mcdu.getDelaySwitchPage();
};

mcdu.onLeftInput[1] = () => {
    CDUProgressPage.ShowReportPage(mcdu);
};
```
| method | ~delay ms | Description |
| --- | --- | --- |
| `getDelaySwitchPage()` | 150-200 | Used for switching pages
| `getDelayBasic()` | 300-400 | Used for basic inputs e.g. alternate airport, ci, fl, temp, constraints, ...
| `getDelayMedium()` | 600-800 | Used for e.g. loading time fore pages
| `getDelayHigh()` | 900-1200 | Used for intense calculation
| `getDelayRouteChange()` | 300-2000+ _(dynamic)_ | Used for changes to the flight plan
| `getDelayFuelPred()` | waypoint count ^ 2 + route lenght | Used for calculation time for fuel pred page
| `getDelayWindLoad()` | waypoint count ^ 2 | Used to load wind data into fms

## References

https://www.youtube.com/watch?v=EoJ9uqMYH_k&start=27

### Advice for testing
Check if given button delay matches expected categorize.
To do so, use mcdu as normal, click through the pages and check if response is good.

Discord username (if different from GitHub): DerL30N#3751

## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created and uploaded.
The build script will have already been run with the latest changes, so no need to rerun it once you download the zip.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the right side, click on the **Artifacts** drop down and click the **A32NX** link

### CODE REVIEW

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_AirportsMonitor.js [hunk 1]
```diff
@@ -212,26 +212,23 @@ class CDUAirportsMonitor {
...
+                    if (value === FMCMainDisplay.clrValue) {
                         this.user_ap = undefined;
-                        mcdu.clearUserInput();
                         // trigger data update next frame
                         this.total_delta_t = update_ival_ms;
                         this.ShowPage(mcdu);
                     } else {
-                        console.log(`"${mcdu.inOut}"`);
+                        console.log(`"${value}"`);
                         console.log(`"${FMCMainDisplay.clrValue}"`);
```
Comment (Benjozork) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_AirportsMonitor.js:224]: Please remove `console` calls for performance.

**Review by Benjozork — CHANGES_REQUESTED** [review]
So the code here seems very repetitive. Can we find a way to avoid that ?

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_AirportsMonitor.js (hunk 1 above)
Comment (MisterChocker) [reply A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_AirportsMonitor.js]: done.

**Comment by MisterChocker** [comment]
> So the code here seems very repetitive. Can we find a way to avoid that ?

Do you mean the central handler I optimized (which removed repetitive code) or the 
```js
mcdu.leftInputDelay[1] = () => {
    return mcdu.getDelaySwitchPage();
};
```

This works the same way we add buttons.
When defining this, I override to the default.
I have chosen the default to be for data insert, so I have created an override for every button that controls page changes only.

**Comment by Benjozork** [comment]
I mean, this exact function is repeated a ton:

```js
() => {
    return mcdu.getDelaySwitchPage();
}
```

You could try putting that function in a constant somewhere ?

**Comment by MisterChocker** [comment]
> I mean, this exact function is repeated a ton:
> 
> ```js
> () => {
>     return mcdu.getDelaySwitchPage();
> }
> ```
> 
> You could try putting that function in a constant somewhere ?

Yes I can do that. How do I define a const this great scope?

**Comment by Benjozork** [comment]
> Yes I can do that. How do I define a const this great scope?

```js
const someNameYouChoose = () => {
    return mcdu.getDelaySwitchPage();
};
```

**Comment by MisterChocker** [comment]
> > Yes I can do that. How do I define a const this great scope?
> 
> ```js
> const someNameYouChoose = () => {
>     return mcdu.getDelaySwitchPage();
> };
> ```

and where sould I do that?

**Comment by Benjozork** [comment]
I guess it's fine for now, I just noticed everything is on different files.

The MCDU will need a restructuring at some point though.

**Review by Benjozork — APPROVED** [review]

**Comment by MisterChocker** [comment]
> Text in the Scratchpad still gets cleared out by system messages such as GPS Primary or Select Desired System on MCDU Menu Page

Should be fixed now.

**Pull request merged**